File: aggregator.py
"""Cross-layer weighted aggregation => direction/confidence/urgency/size_multiplier."""
from __future__ import annotations
from dataclasses import dataclass
from .models import StaircaseProfile, GameState, ForceVector, AuthenticityProfile, RegimeWeights, AggregatedSignal, RegimeType, CupFlipState


@dataclass
class AggregatorConfig:
    streak_floor: float = 0.4
    force_squash_denom: float = 1.0
    auth_center: float = 0.5
    auth_scale: float = 2.0
    confidence_base: float = 0.6
    confidence_auth_weight: float = 0.4
    urgency_pressure_weight: float = 0.5
    urgency_force_weight: float = 0.5
    urgency_force_cap: float = 10.0
    size_base: float = 0.5
    size_scale: float = 1.5
    min_confidence: float = 0.2
    min_authenticity: float = 0.3


class SignalAggregator:
    def __init__(self, config: AggregatorConfig | None = None) -> None:
        self.config = config or AggregatorConfig()

    def aggregate(
        self,
        staircase: StaircaseProfile,
        game_state: GameState,
        force_vector: ForceVector,
        authenticity: AuthenticityProfile,
        regime_weights: RegimeWeights,
        timestamp_ms: int,
    ) -> AggregatedSignal:
        cfg = self.config

        # L1 direction from staircase imbalance
        l1 = staircase.imbalance_ratio

        # L2 direction from pressure/state
        l2 = game_state.pressure
        if game_state.state == CupFlipState.BULL_STREAK:
            l2 = max(l2, cfg.streak_floor)
        elif game_state.state == CupFlipState.BEAR_STREAK:
            l2 = min(l2, -cfg.streak_floor)

        # L3 direction from spectral force (squash)
        l3 = force_vector.total_force
        l3 = l3 / (abs(l3) + cfg.force_squash_denom) if l3 != 0 else 0.0

        # L4 authenticity penalty contributes directional neutrality; low authenticity reduces conviction
        l4 = (authenticity.authenticity_score - cfg.auth_center) * cfg.auth_scale  # [-1,1]

        direction = (
            regime_weights.l1_weight * l1 +
            regime_weights.l2_weight * l2 +
            regime_weights.l3_weight * l3 +
            regime_weights.l4_weight * l4
        )
        direction = max(-1.0, min(1.0, direction))

        raw_conviction = abs(direction)
        confidence = max(0.0, min(1.0, raw_conviction * (cfg.confidence_base + cfg.confidence_auth_weight * authenticity.authenticity_score)))
        urgency = max(0.0, min(1.0, cfg.urgency_pressure_weight * abs(game_state.pressure) + cfg.urgency_force_weight * min(1.0, abs(force_vector.total_force) / cfg.urgency_force_cap)))

        size_multiplier = 0.0 if regime_weights.abstain else (cfg.size_base + cfg.size_scale * confidence * authenticity.authenticity_score)

        abstain = regime_weights.abstain or confidence < cfg.min_confidence 

        return AggregatedSignal(
            direction=direction,
            confidence=confidence,
            urgency=urgency,
            size_multiplier=0.0 if abstain else size_multiplier,
            regime=regime_weights.regime,
            abstain=abstain,
            components={"L1": l1, "L2": l2, "L3": l3, "L4": l4},
            timestamp_ms=timestamp_ms,
        )

File: core/src/aggregator.cpp
#include "p6/pipeline.hpp"
#include <algorithm>
#include <cmath>

namespace p6 {

// ── RegimeClassifier ──────────────────────────────────────────────────────────
// Assigns per-regime layer weights.  Weights always sum to 1.0.
//   TRENDING  : L2 (cup flip) and L3 (spectral force) dominate — momentum play
//   RANGING   : L1 (staircase) and L4 (spoof) dominate — mean-reversion / fake-out
//   VOLATILE  : L3 and L4 dominate — risk / deception aware
//   UNKNOWN   : equal 0.25 each

RegimeWeights RegimeClassifier::classify(RegimeType hint) {
    RegimeWeights rw;
    rw.regime = hint;

    switch (hint) {
        case RegimeType::TRENDING:
            rw.l1_weight = 0.15;
            rw.l2_weight = 0.35;
            rw.l3_weight = 0.35;
            rw.l4_weight = 0.15;
            rw.abstain   = false;
            break;
        case RegimeType::RANGING:
            rw.l1_weight = 0.35;
            rw.l2_weight = 0.15;
            rw.l3_weight = 0.15;
            rw.l4_weight = 0.35;
            rw.abstain   = false;
            break;
        case RegimeType::VOLATILE:
            rw.l1_weight = 0.15;
            rw.l2_weight = 0.20;
            rw.l3_weight = 0.30;
            rw.l4_weight = 0.35;
            rw.abstain   = false;
            break;
        case RegimeType::UNKNOWN:
        default:
            rw.l1_weight = 0.25;
            rw.l2_weight = 0.25;
            rw.l3_weight = 0.25;
            rw.l4_weight = 0.25;
            rw.abstain   = false;
            break;
    }
    return rw;
}

// ── SignalAggregator ──────────────────────────────────────────────────────────
// Matches Python aggregator.py formulas exactly:
//
//   L1 direction = imbalance_ratio
//   L2 direction = pressure, floored to ±streak_floor when in BULL/BEAR_STREAK
//   L3 direction = total_force / (|total_force| + force_squash_denom)
//   L4 direction = (auth_score - auth_center) * auth_scale   clamped [-1,1]
//
//   direction = Σ weight_i * comp_i   clamped [-1, 1]
//
//   confidence = |raw_direction| * (confidence_base + confidence_auth_weight * auth_score)
//   urgency    = pressure_weight * |pressure| + force_weight * min(1, |force|/force_cap)
//   size_mult  = size_base + size_scale * confidence * auth_score  (or 0 if abstain)
//   abstain    = regime.abstain OR confidence < min_confidence OR auth < min_authenticity

SignalAggregator::SignalAggregator() : config_(Config{}) {}
SignalAggregator::SignalAggregator(Config config) : config_(config) {}

Signal SignalAggregator::aggregate(const StaircaseProfile& staircase,
                                   const GameState& game_state,
                                   const ForceVector& force,
                                   const AuthenticityProfile& auth,
                                   const RegimeWeights& regime,
                                   int64_t timestamp_ms) {
    Signal sig;
    sig.timestamp_ms = timestamp_ms;
    sig.regime       = regime.regime;

    // L1: staircase imbalance ratio is already in [-1, 1]
    double l1 = staircase.imbalance_ratio;

    // L2: pressure with streak floor
    double l2 = game_state.pressure;
    if (game_state.state == CupFlipState::BULL_STREAK)
        l2 = std::max(l2, config_.streak_floor);
    else if (game_state.state == CupFlipState::BEAR_STREAK)
        l2 = std::min(l2, -config_.streak_floor);

    // L3: squashed force (soft-sign via x/(|x|+d))
    double fv   = force.total_force;
    double l3   = fv / (std::abs(fv) + config_.force_squash_denom);

    // L4: authenticity mapped to direction
    double l4 = std::clamp(
        (auth.authenticity_score - config_.auth_center) * config_.auth_scale,
        -1.0, 1.0);

    sig.l1_component = l1;
    sig.l2_component = l2;
    sig.l3_component = l3;
    sig.l4_component = l4;

    // Weighted direction
    double raw = regime.l1_weight * l1
               + regime.l2_weight * l2
               + regime.l3_weight * l3
               + regime.l4_weight * l4;

    sig.direction = std::clamp(raw, -1.0, 1.0);

    // Confidence: conviction (|raw|) × (base + auth_weight × auth_score)
    double conviction = std::abs(raw);
    sig.confidence = std::clamp(
        conviction * (config_.confidence_base +
                      config_.confidence_auth_weight * auth.authenticity_score),
        0.0, 1.0);

    // Urgency: pressure term + spectral force term
    sig.urgency = std::clamp(
        config_.urgency_pressure_weight * std::abs(game_state.pressure) +
        config_.urgency_force_weight    * std::min(1.0, std::abs(fv) / config_.urgency_force_cap),
        0.0, 1.0);

    // Abstain logic
    bool should_abstain = regime.abstain
                       || sig.confidence < config_.min_confidence
                       || auth.authenticity_score < config_.min_authenticity;
    sig.abstain = should_abstain;
    sig.size_multiplier = should_abstain ? 0.0
        : config_.size_base + config_.size_scale * sig.confidence * auth.authenticity_score;

    return sig;
}

} // namespace p6
